### PR TITLE
feat(k8sprocessor): support metadata enrichment for  source

### PR DESCRIPTION
**Pull request by sumo-drosiek**
**Description:**

Add metadata enrichment for `pod_name.namespace_name` values

**Link to tracking Issue:** #4309 

**Testing:** unit tests (in progress)

**Documentation:**
inline comments and README update

### CODE REVIEW

**Comment by sumo-drosiek** [comment]
`pod_name.namespace_name` have to be build outside of processor

File: processor/k8sprocessor/doc.go [hunk 1]
```diff
@@ -22,7 +22,7 @@
 // Each rule is specified as a pair of from (representing the rule type) and name (representing the extracted key name).
 // Following rule types are available:
 //   from: "resource_attribute" - allows to specify the attribute name to lookup up in the list of attributes of the received Resource. The specified attribute, if it is present, identifies the Pod that is represented by the Resource.
-//     (the value can contain either IP address or Pod UID)
+//     (the value can contain either IP address, Pod UID or be in  `pod_name.namespace_name` format)
```
Comment (sumo-drosiek) [inline processor/k8sprocessor/doc.go:25]:
```suggestion
//     (the value can contain either IP address, Pod UID or be in `pod_name.namespace_name` format)
```

**Review by Aneurysm9 — DISMISSED** [review]

**Comment by tigrannajaryan** [comment]
@owais please review.

**Comment by sumo-drosiek** [comment]
I added new `pod_association`, because I don't see any processor which could concatenate multiple attributes into one.
`pod_association` `build_hostname` will build hostname from `k8s.pod.name` and `k8s.namespace.name` attributes. I want to avoid changes in `podAssociation` struct.

@dashpole could you review please

**Comment by owais** [comment]
Instead of adding this as a very special case, can we add a bit more general purpose pod associate rule? Perhaps something like:

```
pod_association:
      - from: concatinating
         sources:
             - from: resource_attribute
                name: k8s.pod.name
             - from: resource_attribute
                name: k8s.namespace.name
```

Not proposing this as a final form but just throwing ideas out there. We can also change `pod_association` config completely to better accomodate for such cases. Happy to hear other thoughts. 
Just throwing ideas out ther

**Comment by sumo-drosiek** [comment]
It would work. In addition, the cache keys could be build with the same conditions

**Comment by sumo-drosiek** [comment]
I would propose to use following structure for `pod_association`:

```
pod_association:
  - delimiter: <string>  # delimiter to concatenate attributes from sources. `.` by default
    name: <string>  # resource attribute name when result of concatenation should be put. leave empty to skip
    sources:  # list of attributes for concatenation 
      - from: <string>  # source of an attribute (can be `connection` or `resource_attribute`)
        name: <string>  # name of the attribute in case of using `resource_attribute`
```

example:

```
pod_association:
    - delimiter: "."
      name: hostname
      sources:
        - from: resource_attribute
          name: k8s.pod.name
        - from: resource_atrtribute
          name: k8s.namespace.name
        - from: connection
    - name: pod_uid
      sources:
        - from: resource_attribute
          name: k8s.pod.uid
    - name: k8s.pod.ip
      sources:
        - from: connection
          name: ip
```

---

it would be nice to move `namespace` and `name` from `Pod` struct into `attributes` to make code more general
I would even move most of the fields to attributes

from 
```
// Pod represents a kubernetes pod.
type Pod struct {
	Name       string
	Address    string
	PodUID     string
	Attributes map[string]string
	StartTime  *metav1.Time
	Ignore     bool
	Namespace  string

	DeletedAt time.Time
}
```

to 
```
// Pod represents a kubernetes pod.
type Pod struct {
	# Name       string => attributes[k8s.pod.name]
	# Address    string => attributes[k8s.pod.ip]
	# PodUID     string => attributes[k8s.pod.uid]
	Attributes map[string]string
	StartTime  *metav1.Time
	Ignore     bool
	# Namespace  string => attributes[k8s.namespace.name]

	DeletedAt time.Time
}
```

**Comment by hossain-rayhan** [comment]
Hi @sumo-drosiek, I do have the exact same requirement. I am using `prometheusreceiver` to scrape metrics using kubernetes service discovery where I have only pod_name and namespace. However, in our case these pod_name and namespace are coming as metric labels which are now known as dataPoint attributes. 

So, I was also expecting support for pulling values from metric labels besides resource attributes while generating pod_associations. 

Also wondering, how likely it is that pod_name and namespace will alsways come from resource attributes?

Thanks. 

cc: @dashpole @owais

**Comment by sumo-drosiek** [comment]
@hossain-rayhan 
Regarding dataPoint attributes, we could add support for `metrics` to [groupbyattrsprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/groupbyattrsprocessor). I believe this would be better solution in terms of overall performance and for me `k8sprocessor` should stay at resource attributes level.

**Comment by hossain-rayhan** [comment]
@sumo-drosiek 
Yea, that might be an approach. But for our case, the exporter expects these values as metric labels. If we want to follow this approach, then we need to do two extra conversions-

1.  metric label to resource attributes in a processor.
1.  resource attributes to metric labels in the exporter.

I believe this will technically hamper our performance. Not sure which one would be the best call.

**Comment by sumo-drosiek** [comment]
@hossain-rayhan Is it complicated to modify exporter behavior to look at the resource attributes as well?

**Comment by sumo-drosiek** [comment]
>  Let's discuss this separately if you feel strongly about it.

@owais, Do you mean separate issue, PR?

**Comment by tigrannajaryan** [comment]
@owais are your comments/concerns addressed?

**Comment by sumo-drosiek** [comment]
I would like to add more tests and update `doc.go`

**Comment by owais** [comment]
@dmitryax can you review this as well please?

**Comment by owais** [comment]
Took a quick look. Overall looks good. I'll review in more detail later today or tomorrow. @pmm-sumo @dmitryax please take  a look as well when you can.

Q: The upcoming release notes should mention this as a breaking change and  explain how to migrate to the new config format. Do we have a process in place that allows PRs to commit such notes somewhere so release eng doesn't forget about it?

**Comment by pmm-sumo** [comment]
> Q: The upcoming release notes should mention this as a breaking change and explain how to migrate to the new config format. Do we have a process in place that allows PRs to commit such notes somewhere so release eng doesn't forget about it?

Looking at previous cases like that, I would lean towards adding a compatibility layer which would consume the old config (so keep "From" entry) and print warning message that it is deprecated. WDYT @sumo-drosiek ?

**Comment by sumo-drosiek** [comment]
> > Q: The upcoming release notes should mention this as a breaking change and explain how to migrate to the new config format. Do we have a process in place that allows PRs to commit such notes somewhere so release eng doesn't forget about it?
> 
> Looking at previous cases like that, I would lean towards adding a compatibility layer which would consume the old config (so keep "From" entry) and print warning message that it is deprecated. WDYT @sumo-drosiek ?

Shouldn't complicate the PR, so I can do it. WDYT @owais @bogdandrutu?

**Review by dashpole — APPROVED** [review]
lgtm after nits

**Pull request closed without merging**